### ISSUE TITLE
Failed to generate netconf-config-change notification, but changes were applied

### ISSUE THREAD

**Issue by hi-Mr-ye** [sysrepo/sysrepo#3292]
Hi
I use the function sr_apply_changes (session, 2000) to display error messages:
Waiting for subscription of "ietf-netconf-notifications" failed, previous event "notif" ID 4 was not processed.
Failed to generate netconf-config-change notification, but changes were applied.

May I ask the reason for the error and how to resolve it? the function sr_apply_changes return value is 13.
When the sr_apply_changes function is called, another thread is simultaneously calling the sr_notif_send (session,...., 2000, 0) function. and sr_notif_send return value is 13.
Why is the timeout set to 2000 for the sr-apply_changes function and sr-notif_send function, but it actually takes 10000ms to return the function result？
SR_NOTIF_CB_TIMEOUT and SR_CHANGE_CB_TIMEOUT is 10000

**Comment by michalvasko** [comment]
The problem has occurred before, exactly what the error says, **previous event** was not processed. So look at the previous logs.

**Comment by hi-Mr-ye** [comment]
Thank you for your answer.
I have another question:
Why is the timeout set to 2000 for the sr_apply_changes function and sr_notif_send function, but it actually takes 10000ms to return the function result？
SR_NOTIF_CB_TIMEOUT and SR_CHANGE_CB_TIMEOUT is 10000

**Comment by michalvasko** [comment]
> Why is the timeout set to 2000 for the sr_apply_changes function

Where is this timeout set? `SR_NOTIF_CB_TIMEOUT` is 2000 and `SR_CHANGE_CB_TIMEOUT` is 5000, which are the default timeouts in case users do not specify their own.

**Comment by hi-Mr-ye** [comment]
I use rc = sr_apply_changes(session, 2000) and rc =  sr_notif_send(session, ..., 2000, 0);  but it actually takes 10000ms to return the function result when  display error messages "Failed to generate netconf-config-change notification, but changes were applied"
I have modified the values of SR-NOTIF-CB-TIMEOUT and SR-CHANGE-CB-TIMEOUT to 10000.
![image](https://github.com/sysrepo/sysrepo/assets/90752673/ae3f2c9e-f82b-4580-b458-894a59ef3da1)

**Comment by michalvasko** [comment]
> I have modified the values of SR-NOTIF-CB-TIMEOUT and SR-CHANGE-CB-TIMEOUT to 10000.

Why?

In any case, your observation is correct because, in most cases, the function timeout does not determine the maximum time the called function will block but rather the timeout of a particular set of callbacks that are expected to take by far the largest portion of the whole time the function takes. It is expected it will be almost equal to the whole function timeout in case everything works as expected. In your case there is a major problem with notification processing which is the root cause of the large timeout.

**Comment by hi-Mr-ye** [comment]
Because SR_CHANGE_CB_TIMEOUT=5000 is not enough for me, it has been modified to 10000. Will this cause any unknown problems？



> In your case there is a major problem with notification processing which is the root cause of the large timeout.

How to print out the specific information of this main issue?
I use the sr_log_set_cb function to display error messages:
“Waiting for subscription of "ietf-netconf-notifications" failed, previous event "notif" ID 4 was not processed.
Failed to generate netconf-config-change notification, but changes were applied.”
There was no error message before this.

**Comment by michalvasko** [comment]
> Because SR_CHANGE_CB_TIMEOUT=5000 is not enough for me, it has been modified to 10000. Will this cause any unknown problems？

No, but you can always adjust the timeout for the specific call, it should not be needed to increase it globally. When using *netopeer2-server*, just add the parameter `-t 10` and you get the same behavior.

> How to print out the specific information of this main issue?

Then another process is causing the issue, probably *netopeer2-server*, so look at its output when run with `-d -v3`.

**Comment by hi-Mr-ye** [comment]
Thank you for your answer.
